addons/stock/models/stock_rule.py

```diff
@@ -167,6 +172,21 @@ def _push_prepare_move_copy_values(self, move_to_copy, new_date):
         }
         return new_move_vals
 
+    def _get_procure_method(self, product, qty_needed, location):
+        """ When the procure_method of the rule is 'mts_then_mto', we need to get the quantity available to know if
+        we have to apply MTS or MTO on the moves of the procurement.
+        :param product: The product of the move
+        :param qty_needed: The amount needed of that product
+        :param location: The location source of the product
+        :return: The procure_method to apply on the move
+        """
+        if self.procure_method == 'mts_then_mto':
+            qty_available = product.with_context(location=location.id).virtual_available
```

Comment:
can't you batch virtual_available? read it for all products where the rule is mts_then_mto and pass the values there